torch/compiler/constraint.py

```diff
@@ -0,0 +1,108 @@
+import dataclasses
+import weakref
+
+from typing import Optional
+
+import sympy
+
+import torch
+
+from torch.fx.experimental.symbolic_shapes import StrictMinMaxConstraint
+from torch.utils._sympy.value_ranges import ValueRanges
+
+
+@dataclasses.dataclass
+class ConstraintTarget:
+    """
+    This represents input tensor dimensions.  Don't create this
+    class directly; instead, use :func:`torch._export.dynamic_dim`.
+    """
+
+    w_tensor: weakref.ReferenceType
+    # TODO: We don't need t_id; we can get it off of w_tensor
+    t_id: int
+    dim: int
+
+
+@dataclasses.dataclass
+class Constraint(ConstraintTarget):
```

Comment:
While the export API expects `Constraint`, we are in a funny place because the constructor should not be public: we explicitly ask that you use `dynamic_dim` and operator overloading.

Perhaps we should decouple the class and the type via a type alias? So the type alias is public, but the class is private...